tracks/rust/exercises/bracket-push/mentoring.md

```diff
@@ -1,20 +1,58 @@
 ### Concepts
 
 - stack or recursion
+- `HashMap`s and `HashSet`s
 
 ### Reasonable solutions
 
 A reasonable solution should do the following:
 
+- Keep track of the associations between specific brackets using a `HashMap`, a closure, or something of the like. 
+- Perform as few explicit checks for specific brackets as possible; instead, one possible strategy is to store openers and closers in separate `HashSet`s, and then check if the current char is contained in either set. 
```

Comment:
This is not good advice. A minimal objection: it is more efficient to match each input character exactly once. If we're storing distinct hash sets, we need to perform at least two comparisons for each input.

A longer objection: it's not clear why it would be bad practice to perform explicit checks. I submit that the following [implementation](https://github.com/coriolinus/exercism-rs/blob/master/bracket-push/src/lib.rs) is easier to read and more efficient as the new example; match statements are faster than hash lookups:

```rust
    pub fn brackets_are_balanced(input: &str) -> bool {
        let mut bracket_stack = Vec::new();
        for ch in input.chars() {
            use BracketTypes::*;
            match ch {
                '(' => bracket_stack.push(Parenthesis),
                '[' => bracket_stack.push(Square),
                '{' => bracket_stack.push(Curly),
                ')' => {
                    if bracket_stack.pop() != Some(Parenthesis) {
                        return false;
                    }
                }
                ']' => {
                    if bracket_stack.pop() != Some(Square) {
                        return false;
                    }
                }
                '}' => {
                    if bracket_stack.pop() != Some(Curly) {
                        return false;
                    }
                }
                _ => {}
            }
        }
        bracket_stack.len() == 0
    }
```